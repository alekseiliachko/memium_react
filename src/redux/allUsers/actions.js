import AccountOpenController from "../../api/AccountOpenController";

export const SET_USER_DATA = "SET_USER_DATA";
export const setUserData = (userId, userData) => ({
  type: SET_USER_DATA,
  userId,
  userData,
});

export const fetchUserData = (userId) => async (dispatch) => {
  const res = await AccountOpenController.getAccountsDetails(userId);
  dispatch(setUserData(userId, res.data));
};

export const SET_USER_AVATAR = "SET_USER_AVATAR";
export const setUserAvatar = (userId, avatar) => ({
  type: SET_USER_AVATAR,
  userId,
  avatar,
});
export const fetchUserAvatar = (userId) => async (dispatch) => {
  const res = await AccountOpenController.getAccountsAvatar(userId);
  dispatch(setUserAvatar(userId, res));
};
