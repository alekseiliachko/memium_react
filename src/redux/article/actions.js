import ArticlesController from "../../api/ArticlesController";

export const SET_ARTICLE_DATA = "SET_ARTICLE_DATA";

export const setArticleData = (articleId, articleData) => ({
  type: SET_ARTICLE_DATA,
  articleId,
  articleData,
});

export const fetchArticleData = (articleId) => async (dispatch) => {
  const res = await ArticlesController.getArticleById(articleId);
  dispatch(setArticleData(articleId, res.data));
};
