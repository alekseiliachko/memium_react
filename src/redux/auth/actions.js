import AuthController from "../../api/AuthController";

export const LOGIN_ATTEMPT = "LOGIN_ATTEMPT";
export const LOGIN_SUCCESS = "LOGIN_SUCCESS";
export const LOG_OUT = "LOG_OUT";

export const SIGNUP_ATTEMPT = "SIGNUP_ATTEMPT";
export const SIGNUP_SUCCESS = "SIGNUP_SUCCESS";

export const AUTH_FAILURE = "AUTH_FAILURE";

export const setAuthStatusAttempt = () => ({
  type: LOGIN_ATTEMPT,
});

export const setAuthStatusSuccess = (username) => ({
  type: LOGIN_SUCCESS,
  payload: username,
});

export const logout = () => async (dispatch) => {
  localStorage.removeItem("memium_token");
  localStorage.removeItem("name");
  return {
    type: LOG_OUT,
  };
};

export const authFailure = (err) => ({
  type: AUTH_FAILURE,
  payload: err,
});

export const signupSuccess = () => ({
  type: SIGNUP_SUCCESS,
});

export const restoreSessionAttempt = () => (dispatch) => {
  dispatch(setAuthStatusAttempt());
  if (localStorage.memium_token) {
    AuthController.updateAuthHeader(localStorage.memium_token);
    dispatch(setAuthStatusSuccess(localStorage.name));
  } else {
    dispatch(logout());
  }
};

export const loginAttempt = (userInfo) => async (dispatch) => {
  dispatch(setAuthStatusAttempt());
  const res = await AuthController.login(userInfo);
  AuthController.updateAuthHeader(res.data.token);
  localStorage.setItem("name", res.data.username);
  dispatch(setAuthStatusSuccess(res.data.username));
};

export const signupAttempt = (userInfo) => async (dispatch) => {
  const res = AuthController.signup(userInfo);
  dispatch(signupSuccess());
};
