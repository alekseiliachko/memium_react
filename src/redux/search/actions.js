import AccountsController from "../../api/AccountsController";

export const UPDATE_SEARCH_RESULTS = "UPDATE_SEARCH_RESULTS";
export const updateSearchResults = (articles, authors) => ({
  type: UPDATE_SEARCH_RESULTS,
  articles,
  authors,
});

export const fetchSearchResults = (search) => async (dispatch) => {
  const res = await AccountsController.getArticlesBySearch(search);
  dispatch(updateSearchResults(res.data.articles, res.data.accounts));
};
