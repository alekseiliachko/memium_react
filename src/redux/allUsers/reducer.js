import { SET_USER_AVATAR, SET_USER_DATA } from "./actions";

export const initialState = {
  userDataById: {},
};

export const allUsersReducer = (store = initialState, action) => {
  switch (action.type) {
    case SET_USER_DATA:
      return {
        ...store,
        userDataById: {
          ...store.userDataById,
          [action.userId]: {
            ...store.userDataById[action.userId],
            ...action.userData,
          },
        },
      };
    case SET_USER_AVATAR:
      return {
        ...store,
        userDataById: {
          ...store.userDataById,
          [action.userId]: {
            ...store.userDataById[action.userId],
            avatar: action.avatar,
          },
        },
      };
    default:
      return store;
  }
};
