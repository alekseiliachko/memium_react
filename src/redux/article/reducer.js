import { RESET_ARTICLE_DATA, SET_ARTICLE_DATA } from "./actions";

const initialState = {
  articleDataById: {},
};

export const articleReducer = (store = initialState, action) => {
  switch (action.type) {
    case SET_ARTICLE_DATA:
      return {
        ...store,
        articleDataById: {
          ...store.articleDataById,
          [action.articleId]: action.articleData,
        },
      };
    case RESET_ARTICLE_DATA:
      return {
        ...store,
        articleDataById: {
          ...store.articleDataById,
          [action.articleId]: undefined,
        },
      };
    default:
      return store;
  }
};
