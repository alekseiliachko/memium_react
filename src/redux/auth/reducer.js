import { LOGIN_ATTEMPT, LOGIN_SUCCESS, LOG_OUT } from "./actions";

export const AUTH_STATE = {
  LOGGED_OUT: "LOGGED_OUT",
  SUCCESS: "SUCCESS",
  ATTEMPT: "ATTEMPT",
};

const initialState = {
  username: "",
  authState: AUTH_STATE.ATTEMPT,
};

export const authReducer = (store = initialState, action) => {
  switch (action.type) {
    case LOGIN_ATTEMPT:
      return {
        ...store,
        authState: AUTH_STATE.ATTEMPT,
      };
    case LOGIN_SUCCESS:
      return {
        ...store,
        username: action.payload,
        authState: AUTH_STATE.SUCCESS,
      };
    case LOG_OUT:
      return {
        ...store,
        username: "",
        authState: AUTH_STATE.LOGGED_OUT,
      };
    default:
      return store;
  }
};
