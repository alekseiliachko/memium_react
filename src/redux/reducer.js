import { combineReducers, createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { authReducer } from "./auth/reducer";
import { userReducer } from "./user/reducer";
import { articleReducer } from "./article/reducer";
import { allUsersReducer } from "./allUsers/reducer";
import { searchReducer } from "./search/reducer";

const rootState = combineReducers({
  authReducer,
  userReducer,
  articleReducer,
  allUsersReducer,
  searchReducer,
});

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const middleware = [thunk];
export function configureStore() {
  const store = createStore(
    rootState,
    composeEnhancers(applyMiddleware(...middleware))
  );
  return store;
}
