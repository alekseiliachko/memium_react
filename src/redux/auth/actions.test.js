import {
  setAuthStatusAttempt,
  setAuthStatusSuccess,
  logout,
  authFailure,
  signupSuccess,
  signupAttempt,
  loginAttempt,
  restoreSessionAttempt,
} from "./actions";
import { AUTH_STATE, initialState } from "./reducer";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
jest.mock("../../api/AuthController");
import AuthController from "../../api/AuthController";
import ArticlesController from "../../api/ArticlesController";
import { fetchArticleData } from "../article/actions";

const mockStore = configureMockStore([thunk]);

describe("actions", () => {
  let store;
  beforeEach(() => {
    store = mockStore(initialState);
  });
  it("should create LOGIN_ATTEMPT action", () => {
    store.dispatch(setAuthStatusAttempt());
    expect(store.getActions()).toMatchSnapshot();
  });
  it("should create AUTH_FAILURE action", () => {
    store.dispatch(authFailure(""));
    expect(store.getActions()).toMatchSnapshot();
  });
  it("should create LOGIN_SUCCESS action", () => {
    store.dispatch(setAuthStatusSuccess(""));
    expect(store.getActions()).toMatchSnapshot();
  });
  it("should create SIGNUP_SUCCESS action", () => {
    store.dispatch(signupSuccess());
    expect(store.getActions()).toMatchSnapshot();
  });
  it("should create LOG_OUT action", () => {
    store.dispatch(logout());
    expect(store.getActions()).toMatchSnapshot();
  });
  it("create LOGIN_ATTEMPT, LOGIN_SUCCESS if token exist", () => {
    localStorage.setItem("memium_token", "token");
    localStorage.setItem("name", "name");
    store.dispatch(restoreSessionAttempt());
    expect(store.getActions()).toMatchSnapshot();
  });
  it("create LOGIN_ATTEMPT, LOGOUT if token not exist", () => {
    localStorage.removeItem("memium_token");
    store.dispatch(restoreSessionAttempt());
    expect(store.getActions()).toMatchSnapshot();
  });
  it("should clear localStorage after logout", () => {
    localStorage.setItem("memium_token", "token");
    localStorage.setItem("name", "name");
    store.dispatch(logout());
    expect(localStorage.getItem("memium_token")).toBeNull();
    expect(localStorage.getItem("name")).toBeNull();
  });
});

describe("async actions", () => {
  let store;
  beforeEach(() => {
    store = mockStore({ initialState });
  });

  it("create SIGNUP_SUCCESS when fetching has been done", () => {
    AuthController.signup.mockResolvedValue();
    return store.dispatch(signupAttempt({})).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
  it("create LOGIN_ATTEMPT, LOGIN_SUCCESS when fetching has been done", () => {
    const user = {
      username: "",
      token: "",
    };
    AuthController.login.mockResolvedValue({ data: user });
    return store.dispatch(loginAttempt({})).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
