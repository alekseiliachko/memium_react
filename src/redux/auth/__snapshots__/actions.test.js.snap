// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions create LOGIN_ATTEMPT, LOGIN_SUCCESS if token exist 1`] = `
Array [
  Object {
    "type": "LOGIN_ATTEMPT",
  },
  Object {
    "payload": "name",
    "type": "LOGIN_SUCCESS",
  },
]
`;

exports[`actions create LOGIN_ATTEMPT, LOGOUT if token not exist 1`] = `
Array [
  Object {
    "type": "LOGIN_ATTEMPT",
  },
  Object {
    "type": "LOG_OUT",
  },
]
`;

exports[`actions should create AUTH_FAILURE action 1`] = `
Array [
  Object {
    "payload": "",
    "type": "AUTH_FAILURE",
  },
]
`;

exports[`actions should create LOG_OUT action 1`] = `
Array [
  Object {
    "type": "LOG_OUT",
  },
]
`;

exports[`actions should create LOGIN_ATTEMPT action 1`] = `
Array [
  Object {
    "type": "LOGIN_ATTEMPT",
  },
]
`;

exports[`actions should create LOGIN_SUCCESS action 1`] = `
Array [
  Object {
    "payload": "",
    "type": "LOGIN_SUCCESS",
  },
]
`;

exports[`actions should create SIGNUP_SUCCESS action 1`] = `
Array [
  Object {
    "type": "SIGNUP_SUCCESS",
  },
]
`;

exports[`async actions create LOGIN_ATTEMPT, LOGIN_SUCCESS when fetching has been done 1`] = `
Array [
  Object {
    "type": "LOGIN_ATTEMPT",
  },
  Object {
    "payload": "",
    "type": "LOGIN_SUCCESS",
  },
]
`;

exports[`async actions create SIGNUP_SUCCESS when fetching has been done 1`] = `
Array [
  Object {
    "type": "SIGNUP_SUCCESS",
  },
]
`;
